### PR TITLE
Bluetooth: Mesh: Extended advertiser as default

### PR DESCRIPTION
**Pull request by Balaklaka**
Setting the extended advertiser as default advertiser to improve both preformence and reliability.

Legacy advertiser can occasionally do more message retransmissions than requested. When using the legacy advertiser, the stack sleeps for at least 50 ms after starting advertising a message, which may result in more messages to be advertised than requested.

Signed-off-by: Ingar Kulbrandstad <ingar.kulbrandstad@nordicsemi.no>

### CODE REVIEW

**Review by alxelax — CHANGES_REQUESTED** [review]
The low latency feature is not compatible with the extended advertiser. You broke bsim test.

File: samples/bluetooth/mesh/prj.conf [hunk 1]
```diff
@@ -21,6 +20,7 @@ CONFIG_BT_PERIPHERAL=y
 CONFIG_BT=y
 CONFIG_BT_TINYCRYPT_ECC=y
 CONFIG_BT_L2CAP_TX_BUF_COUNT=5
+CONFIG_BT_MAX_CONN=2
```
Comment (PavelVPV) [inline samples/bluetooth/mesh/prj.conf:23]: Why was this changed?

File: tests/bluetooth/mesh_shell/prj.conf [hunk 2]
```diff
@@ -22,15 +22,14 @@ CONFIG_BT_PERIPHERAL=y
 CONFIG_BT_TINYCRYPT_ECC=y
 CONFIG_BT_DEVICE_NAME="Zephyr Mesh"
 CONFIG_BT_GATT_CACHING=n
+CONFIG_BT_MAX_CONN=2
```
Comment (PavelVPV) [inline tests/bluetooth/mesh_shell/prj.conf:25]: Why was this changed?

**Review by PavelVPV — CHANGES_REQUESTED** [review]
I tried to collect things that need to be updated (though something may still come up during testing):

1. Please remove `CONFIG_BT_EXT_ADV=y` in mesh bsim tests.
2. Please use legacy adv in:
    - `tests/bluetooth/mesh` where it necessary
    - `samples/bluetooth/mesh/microbit_gatt.conf`
    - `samples/bluetooth/mesh/nrf51_qfaa.conf`
    - `samples/bluetooth/mesh/prj_bbc_microbit.conf`
    - `samples/bluetooth/mesh_demo/prj_bbc_microbit.conf`
    - `samples/bluetooth/mesh_demo/prj_nrf51_blenano.conf`
3. You may need to create a separate `prj_nrf51_blenano.conf` file to select legacy adv for `nrf51_blenano` target in `samples/bluetooth/mesh`.
Same for `nrf51dk_nrf51422` target in `samples/bluetooth/mesh_demo`.
Please check that these samples compile for these targets.
4. You may need to force legacy adv in `samples/boards/reel_board/mesh_badge` sample.
5. Please make sure that the following samples still compiles for `nrf52840dk_nrf52840`:
    - samples/boards/nrf/mesh/onoff-app
    - samples/boards/nrf/mesh/onoff_level_lighting_vnd_app
6. Please add a note to the release notes.
7. Since the reason for the switch is inaccuracy of the legacy adv, I think it worth to add a note about that in `mesh/Kconfig`.

File: subsys/bluetooth/mesh/Kconfig [hunk 3]
```diff
@@ -9,6 +9,7 @@ menuconfig BT_MESH
 	select TINYCRYPT_AES
 	select TINYCRYPT_AES_CMAC
 	select BT_HOST_CCM
+	select BT_EXT_ADV
```
[the remark this answers is not in the document]
Comment (Balaklaka) [reply subsys/bluetooth/mesh/Kconfig:12]: I was first thinking about doing it the way you suggested, but I was trying to avoid changing anything in bluetooth. It is still possible to use BT_MESH_ADV_LEGACY, and turn off BT_CTRL_ADV_EXT, but it will not be possible to turn off BT_EXT_ADV. But I think you are correct and we need to change it this way.

[the remark this answers is not in the document]
Comment (omkar3141) [reply subsys/bluetooth/mesh/Kconfig:12]: `default y if BT_MESH`:
Will this generate compilation warning if someone were to set `BT_EXT_ADV=n` in their prj.conf?

Comment (omkar3141) [inline subsys/bluetooth/mesh/Kconfig:12]: Where do you plan to put the updated defaults for "BT_MESH_RELAY_ADV_SETS, BT_MESH_ADV_EXT_GATT_SEPARATE", downstream?

**Review by omkar3141 — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/Kconfig (hunk 3 above)
[the remark this answers is not in the document]
Comment (Balaklaka) [reply subsys/bluetooth/mesh/Kconfig]: It will not give a warning when BT_EXT_ADV=n is set.

File: samples/bluetooth/mesh/prj.conf (hunk 1 above)
Comment (Balaklaka) [reply samples/bluetooth/mesh/prj.conf:23]: Default this is set to 1, and I found out this need to be 2. It is 1 when CONFIG_BT_OBSERVER=y and CONFIG_BT_PERIPHERAL=y, but 2 if CONFIG_BT_CENTRL=y and CONFIG_BT_PERIPHERAL=y is set.

Comment (omkar3141) [reply samples/bluetooth/mesh/prj.conf:23]: Why would someone set CONFIG_BT_CENTRAL in a mesh sample? It does not look like any functionality in this sample needs BT_CENTRAL.

Comment (alxelax) [reply samples/bluetooth/mesh/prj.conf:23]: due to gatt proxy client?

Comment (PavelVPV) [reply samples/bluetooth/mesh/prj.conf:23]: Yeah, but Central role is not enabled here, neither PB GATT or Proxy Client. Unless I miss something...

File: tests/bluetooth/mesh_shell/prj.conf (hunk 2 above)
Comment (Balaklaka) [reply tests/bluetooth/mesh_shell/prj.conf:25]: Same reason as above.

**Comment by Balaklaka** [comment]
> I tried to collect things that need to be updated (though something may still come up during testing):
> 
> 1. Please remove `CONFIG_BT_EXT_ADV=y` in mesh bsim tests.
> 2. Please use legacy adv in:
>    
>    * `tests/bluetooth/mesh` where it necessary
>    * `samples/bluetooth/mesh/microbit_gatt.conf`
>    * `samples/bluetooth/mesh/nrf51_qfaa.conf`
>    * `samples/bluetooth/mesh/prj_bbc_microbit.conf`
>    * `samples/bluetooth/mesh_demo/prj_bbc_microbit.conf`
>    * `samples/bluetooth/mesh_demo/prj_nrf51_blenano.conf`
> 3. You may need to create a separate `prj_nrf51_blenano.conf` file to select legacy adv for `nrf51_blenano` target in `samples/bluetooth/mesh`.
>    Same for `nrf51dk_nrf51422` target in `samples/bluetooth/mesh_demo`.
>    Please check that these samples compile for these targets.
> 4. You may need to force legacy adv in `samples/boards/reel_board/mesh_badge` sample.
> 5. Please make sure that the following samples still compiles for `nrf52840dk_nrf52840`:
>    
>    * samples/boards/nrf/mesh/onoff-app
>    * samples/boards/nrf/mesh/onoff_level_lighting_vnd_app
> 6. Please add a note to the release notes.
> 7. Since the reason for the switch is inaccuracy of the legacy adv, I think it worth to add a note about that in `mesh/Kconfig`.

1. Done
2. Added to the files.
3. Did not need to create the config file ised the nrf51_qfaa.conf file. Works with the nrf51dk_nrf51422
4. Did not need the legacy adv to compile
5. Done
6. Done
7. Done

File: samples/bluetooth/mesh/prj.conf (hunk 1 above)
Comment (Balaklaka) [reply samples/bluetooth/mesh/prj.conf:23]: PB_GATT is added 5 lines down.

Comment (PavelVPV) [reply samples/bluetooth/mesh/prj.conf:23]: But not client

File: samples/bluetooth/mesh/prj_bbc_microbit.conf [hunk 4]
```diff
@@ -1,5 +1,5 @@
 CONFIG_INIT_STACKS=y
-CONFIG_ISR_STACK_SIZE=768
+CONFIG_ISR_STACK_SIZE=512
```
Comment (alxelax) [inline samples/bluetooth/mesh/prj_bbc_microbit.conf:2]: Is there any confirmation if mesh works on this platform after reducing stack size?

File: samples/bluetooth/mesh/prj.conf (hunk 1 above)
Comment (omkar3141) [reply samples/bluetooth/mesh/prj.conf:23]: Just check if you really need BT_MAX_CONN=2, by flashing and provisioning the sample. The CENTRAL should not be selected, if it is getting selected there is some misconfiguration.

File: subsys/bluetooth/mesh/Kconfig [hunk 5]
```diff
@@ -363,6 +363,9 @@ config BT_MESH_ADV_LEGACY
 	  advertising is significantly slower than the extended advertising, but
 	  is supported by all controllers.
 
+	  WARNING: The legacy advertiser can occasionally do more message
+	  retransmissions than requested.
+
```
Comment (omkar3141) [inline subsys/bluetooth/mesh/Kconfig:368]:
```suggestion
	  WARNING: The legacy advertiser can occasionally do more message
	  retransmissions than requested because of limitations of HCI 
	  interface API.
	  
```

**Review by omkar3141 — CHANGES_REQUESTED** [review]
Just a couple of minor things need to be verified. Otherwise, this looks fine.

File: samples/bluetooth/mesh/prj_bbc_microbit.conf (hunk 4 above)
Comment (Balaklaka) [reply samples/bluetooth/mesh/prj_bbc_microbit.conf]: Reverted it as it was not needed when RELAY was removed.

File: tests/bluetooth/mesh_shell/prj.conf (hunk 2 above)
Comment (Balaklaka) [reply tests/bluetooth/mesh_shell/prj.conf:25]: If this is not set to 2, the mesh_shell will give this error when doing "mesh prov pb-gatt on" -> "<err> bt_mesh_adv_ext: adv_start: Advertising failed: err -12"

File: samples/bluetooth/mesh/prj.conf (hunk 1 above)
Comment (Balaklaka) [reply samples/bluetooth/mesh/prj.conf]: This was not necessary to be 2.

**Review by alxelax — APPROVED** [review]

**Review by LingaoM — APPROVED** [review]

**Review by PavelVPV — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]
